lib/puppet/etc.rb

```
@@ -0,0 +1,126 @@
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrnam(groupname)
+      if group_entry = ::Etc.getgrnam(groupname)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getgrid searches /etc/group file for an entry corresponding to id.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrgid(id)
+      if group_entry = ::Etc.getgrgid(id)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getpwuid searches /etc/passwd file for an entry corresponding to id.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwuid(id)
+      if user_entry = ::Etc.getpwuid(id)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    private
+    # Utility method for converting the String values of a struct returned by
+    # the Etc module to UTF-8. Structs returned by the ruby Etc module contain
+    # members with fields of type String, Integer, or Array of Strings, so we
+    # handle these types. Otherwise ignore fields.
+    #
+    # NOTE: If a string cannot be converted to UTF-8, this leaves the original
+    # string string intact in the Struct.
+    #
+    # Warning! This is a destructive method - the struct passed is modified!
+    #
+    # @api private
+    # @param [Etc::Passwd or Etc::Group struct]
+    # @return [Etc::Passwd or Etc::Group struct] the original struct with values
+    #   converted to UTF-8 if possible, or the original value intact if not
+    def convert_field_values_to_utf8!(struct)
+      struct.each_with_index do |value, index|
+        if value.is_a?(String)
+          struct[index] = Puppet::Util::CharacterEncoding.convert_to_utf_8(value) rescue value
```

Or maybe you mean that we should be surfacing that error here? I intended for Puppet::Util::CharacterEncoding::convert_to_utf_8 to raise a Puppet::Error if it could not convert a string to UTF-8 and allow users to handle that error as they saw fit - in this case, per feedback we're hiding the failure to convert and just continuing?